Instructor: home page: show response published status separate from submission status #4536

HirdayGupta opened the pull request:
Fixes #4536

This one turned out to be bigger than expected!

Note: I have updated the InstructorHomePage, StudentHomePage, and InstructorFeedbackSessionsPage in this PR with the new table columns.

### Outline of the solution:

**To add the new status columns (`Submissions` and `Published?`) to the various tables:**

- Added the `publishedStatus` data attribute (private data field, initialization via constructor and getter method) to the required classes, FeedbackSessionsTableRow (for the Instructor Sessions Page) and HomeFeedbackSessionRow (for the Instructor and Student home pages). Modified the original `status` variables in these classes to `submissionStatus`.
- Update the constructor calls for these classes (Which are in the *PageData classes).
    - In order to so, I first had to write utility methods to get the correct value that I wanted to inject into the cell. This involved editing the present method to remove any `publishedStatus` related checks from it (renamed accordingly) and adding a new method that returns either "yes/no" for the `Published?` column.
    - Then called these utility methods in the constructors as required.
- Finally, update the `instructor/feedbacks/feedbackSessionsTable.tag`, `instructor/home/courseTable.tag` and `student/home/courseTable.tag` tag files to add in the new data attributes.

**To Update the Tooltips:**
- Followed a similar process as above for the tooltips that involved adding the new `publishedTooltip` data attribute and editing the current 'tooltip' attribute to `submissionsTooltip` ==> writing new methods to populate the new tooltip and editing the existing one to remove any `published` data from it ==> and then adding the new tooltips to the required tag files.
    - The only difference here was that I had to add in the new tooltips in `Const.Tooltips`

**Testing:**
- Updated the `StudentHomePageDataTest` (see 2aa06ff). Managed to locate this test as I had changed the method names of some of the methods it called. Proceeded to not only fix the compilation errors but also add checks for tooltip and table cell content. 
    - Was unable to find a similar testing method for the Instructor Home Page and the `InstructorFeedbackSessionsPageDataTest` did not seem to test the cell contents and tooltips as the `StudentHomePageDataTest` did. Anyway, the test I added in `StudentHomePageDataTest` seemed to cover most of the permutation and combinations.
- Updated all the expected HTML using god mode. I feel this expected HTML change is enough verification for this change and that this combined with the test I added above should suffice as testing for this PR. Let me know if you disagree.

@damithc Sir if you could please review the following portion of the PR description and see if anything is different from the behavior you intended:

### The new behavior:

**Instructors:**

- `Submissions` Column:
    - Possible Values: `Open`, `Closed`, `Awaiting`, `Private`
    - Tooltips: Unchanged from the original tooltips associated with these statusses, except that they no longer contain  `published?` data.
- `Published?` Column:
    - Possible Values: `Yes`, `No`
    - Tooltips: `The responses for this session are visible.`, `The responses for this session are not visible.` And a **special case** wherein if the session is private, the `No` is explained with the following tooltip: `This feedback session is not published as it is private and only visible to you.` @damithc let me know if this sounds acceptable to you.

**Students:**
- `Submissions` Column:
    - Possible Values: `Submitted`, `Pending`, `Awaiting`, `Closed`.
    - Tooltips: Unchanged from the original tooltips associated with these statusses, except that they no longer contain  `published?` data.
- `Published?` Column:
    - Possible Values: `Yes`, `No`
    - Tooltips: `The responses for the session have been published and can now be viewed.`, `The responses for the session have not yet been published and cannot be viewed.`

HirdayGupta commented:
**InstructorHomePage:**

![yltx0jfc8k](https://user-images.githubusercontent.com/18088721/29489031-c4b7ccb4-8534-11e7-8ec8-72d00881a051.gif)

**InstructorFeedbackSessionsPage:** (Tooltips same as above)

<img width="814" alt="screen shot 2017-08-19 at 11 19 29 pm" src="https://user-images.githubusercontent.com/18088721/29489042-e93367e2-8534-11e7-922f-ac30bcf94afc.png">

**StudentHomePage** (Ignore the session names 😅)

![8tjgejany3](https://user-images.githubusercontent.com/18088721/29489075-80270af0-8535-11e7-91ad-92f60dc95e40.gif)

HirdayGupta commented:
Ready for review.

damithc commented:
1. After seeing the UI, I think `Published?` could confuse instructors into thinking that it is whether session (not results) is visible to students. This could mislead them to publishing the session prematuarly.
Perhaps we should use `Responses: Published, Not Published`?

2. If the session settings says this session will `Never` be published, we should show `-` in that column?

HirdayGupta commented:
@damithc Ready for review sir. Just to be clear: the changes you recommended in the first point are only for the instructor's pages, correct?

**Summary of changes:**
- Changed the heading of the table to `Responses` for instructors only.
- Changed the cell content to display `Published/Not Published` for instructors only.
- As per your second suggestion, set the `publishedStatus` to display` "-"` if resultsVisibleFrom is set to NEVER.
- Added new tooltips to clarify why the "-" is being displayed.

damithc commented:
Good work @HirdayGupta 
I'll wait for a senior dev to review this.

chowyb commented on src/test/java/teammates/test/cases/pagedata/StudentHomePageDataTest.java:
```diff
@@ -107,26 +110,31 @@ private void testOldCourseTable(CourseDetailsBundle oldCourse, CourseTable cours
                                 + Const.Tooltips.STUDENT_FEEDBACK_SESSION_STATUS_CLOSED,
-                            "Closed");
+                            Const.Tooltips.STUDENT_FEEDBACK_SESSION_STATUS_NOT_PUBLISHED,
+                            "Closed", "No");
```
Are these "Yes" and "No"s a carryover from the previous implementation?

HirdayGupta replied:
@chowyb I'm not sure I fully understand your question. The "yes" and "no" is just to verify that the correct status is being displayed in the new column I added in this PR, depending on the status of the session being checked in the given test case.

chowyb replied:
@HirdayGupta yep upon looking at my comment again I can see why 😅 I was just wondering why the strings "Yes" and "No" were used, since looking at the more recent implementation, the strings being used for publishing seem to be "Published" and "Not Published" instead, while the original change in this PR seems to have used "Yes" and "No" previously.

HirdayGupta replied:
@chowyb Ahh, I undertand your confusion! 😅 

If look closely, this is the `StudentHomePageDataTest`. As per @damithc's [suggestion](https://github.com/TEAMMATES/teammates/pull/7965#issuecomment-323556782), I had changed it to `published/not published` only for the Instructor Pages. Student Home Page still looks the same as the screenshot posted under the StudentHomePage heading [here](https://github.com/TEAMMATES/teammates/pull/7965#issuecomment-323538063). Hence, this test still works.

damithc commented:
@HirdayGupta even for student pages, we can use `Responses` `Published` `Not Published`, right?

HirdayGupta commented:
@damithc yes sir, absolutely we can. In fact that would be better for consistency. Don't know why I thought this would be better :sweat_smile:

Will push the updated code tonight so that @chowyb can do a full review the next morning.

HirdayGupta commented:
@damithc @chowyb Ready for review. Changed from 'Published?: Yes/No` to `Responses: Published/Not Published`.

**New studentHomePageUI:**

<img width="1208" alt="screen shot 2017-08-22 at 9 45 33 pm" src="https://user-images.githubusercontent.com/18088721/29575827-542ac1bc-8783-11e7-8a76-f5fb6ce5ac3c.png">

damithc commented:
Looks good @HirdayGupta Can you check the instructor/student help pages to see if anything there needs updating to match this PR?

HirdayGupta commented:
@damithc Just went through the entire InstructorHelp and StudentHelp pages. I don't think there's anything there that gets affected by the changes here.

HirdayGupta commented:
@damithc @chowyb Any issues with the merge?
